src/CodeGen_Vulkan_Dev.cpp

```diff
@@ -0,0 +1,1304 @@
+        // concatenated to form a complete SPIR-V module.  We also
+        // represent the temporaries as vectors of uint32_t rather than
+        // char for ease of adding words to them.
+        std::vector<uint32_t> spir_v_header;
+        std::vector<uint32_t> spir_v_entrypoints;
+        std::vector<uint32_t> spir_v_execution_modes;
+        std::vector<uint32_t> spir_v_annotations;
+        std::vector<uint32_t> spir_v_types;
+        std::vector<uint32_t> spir_v_kernels;
+        // The next one is cleared in between kernels, and tracks the allocations
+        std::vector<uint32_t> spir_v_kernel_allocations;
+
+        // Id of entry point for kernel currently being compiled.
+        uint32_t current_function_id;
+
+        // Top-level function for adding kernels
+        void add_kernel(const Stmt &s, const std::string &name, const std::vector<DeviceArgument> &args);
+
+        // Function for allocating variables in function scope, with optional initializer.
+        // These will appear at the beginning of the function, as required by SPIR-V
+        void add_allocation(uint32_t result_type_id, uint32_t result_id, uint32_t storage_class, uint32_t initializer = 0);
+
+        std::map<Type, uint32_t> type_map;
+        std::map<std::pair<Type, uint32_t>, uint32_t> pointer_type_map;
+        std::map<Type, uint32_t> pair_type_map;
+        std::map<std::string, uint32_t> constant_map;
+
+        void add_instruction(std::vector<uint32_t> &region, uint32_t opcode,
+                             std::initializer_list<uint32_t> words);
+        void add_instruction(uint32_t opcode, std::initializer_list<uint32_t> words);
+        void add_instruction(std::vector<uint32_t> &region, uint32_t opcode,
+                             std::vector<uint32_t> words);
+        void add_instruction(uint32_t opcode, std::vector<uint32_t> words);
+        uint32_t map_type(const Type &type);
+        uint32_t map_pointer_type(const Type &type, uint32_t storage_class);
+        uint32_t map_type_to_pair(const Type &t);
+        uint32_t emit_constant(const Type &t, const void *data);
+        void scalarize(const Expr &e);
+
+        // The scope contains both the symbol and its storage class
+        Scope<std::pair<uint32_t, uint32_t>> symbol_table;
+
+        // The workgroup size.  Must be the same for all kernels.
+        uint32_t workgroup_size[3];
+
+        struct PhiNodeInputs {
+            uint32_t ids[4];
+        };
+        // Returns Phi node inputs.
+        template<typename StmtOrExpr>
+        PhiNodeInputs emit_if_then_else(const Expr &condition, StmtOrExpr then_case, StmtOrExpr else_case);
+    } emitter;
+
+    std::string current_kernel_name;
+};
+
+// --
+
+void CodeGen_Vulkan_Dev::SPIRVEmitter::add_instruction(std::vector<uint32_t> &region, uint32_t opcode,
+                                                       std::initializer_list<uint32_t> words) {
+    region.push_back(((1 + words.size()) << 16) | opcode);
```

Comment:
> ((1 + words.size()) << 16) | opcode

nit: This formula is repeated several times and seems like it could be error-prone, probably worth an inline wrapper?

Reply:
Agreed ... this will be going away.